compiler/rustc_lexer/src/lib.rs

```diff
@@ -365,6 +370,49 @@ impl Cursor<'_> {
                 _ => self.ident_or_unknown_prefix(),
             },
 
+            // Guarded string literal (reserved syntax).
+            '#' if matches!(self.first(), '"' | '#') => {
+                // Create a backup to restore later if this
+                // turns out to not be a guarded literal.
+                let backup = self.clone();
```

Comment:
So far, this approach has *only* being used for the benefit of diagnostics, but won't this become a language requirement to have this kind of "unlimited" lookahead in all alternative implementations?

Reply:
It depends. Alternative implementations could choose to have fully edition-dependent lexing. In that case, they would not need this backtracking.

Reply:
I also don't like this snapshotting. Should `Cursor` gain an `edition` field? Seems like the right way to do it.

Reply:
A couple questions:

1. Are you okay with adding an edition field to cursor?
2. Should I include that refactoring in this PR,
3. Or should I follow up with a separate PR for that refactoring?